compiler/rustc_next_trait_solver/src/solve/trait_goals.rs

```diff
@@ -78,13 +78,22 @@ where
                 TypingMode::Coherence => Certainty::AMBIGUOUS,
                 TypingMode::Analysis { .. }
                 | TypingMode::Borrowck { .. }
+                | TypingMode::Reflection
                 | TypingMode::PostBorrowckAnalysis { .. }
                 | TypingMode::PostAnalysis => return Err(NoSolution),
             },
 
             // Impl matches polarity
             (ty::ImplPolarity::Positive, ty::PredicatePolarity::Positive)
-            | (ty::ImplPolarity::Negative, ty::PredicatePolarity::Negative) => Certainty::Yes,
+            | (ty::ImplPolarity::Negative, ty::PredicatePolarity::Negative) => {
+                if let TypingMode::Reflection = ecx.typing_mode()
+                    && !cx.is_fully_generic_for_reflection(impl_def_id)
+                {
+                    return Err(NoSolution);
```

Comment:
This is way over my head, lol, but just to sanity-check: could this branch be hoisted at the beginning of the function, as in, any `impl` that is not `fully_generic_for_reflection()`, when checking in `Reflection` mode, ought to result in `Err(NoSolution)`, right? As in, it's not _specific_ to this double-`Negative` branch?